### ISSUE TITLE
Support JS compilation

### ISSUE THREAD

**Issue by Whathecode** [cph-cachet/carp.protocols.core#15]
In the future, there will be an opportunity to compile this library to Javascript. This would allow it to be used in React and React Native.

This issue tries to keep track of the current known hurdles which will prevent us from targeting JS at the moment:

- [PolymorphicSerializer does not yet work for JS](https://github.com/Kotlin/kotlinx.serialization/issues/101)
- The library currently relies on reflection here and there which is JVM-specific.

**Comment by Whathecode** [comment]
Although polymorph serialization does not yet work, I found a workaround to at least allow compiling a multiplatform project with common serialization code, in which the JVM target can use serialization, but the JS target does not (it is likely that the same serialization code will support JS serialization in the future). This works by using `expect` and `actual` (Kotlin multiplatform feature), and mapping the the `kotlinx.serialization.Serializable` annotation to a dummy annotation on the JS target. Yet, again, [I had to jump through an additional hurdle to get this to work](https://youtrack.jetbrains.com/issue/KT-25946) (possibly due to current Kotlin `typealias` limitations).

I am now working towards making all tests between both target platforms shared (in `common`). Since the JavaScript target does not support serialization quite a few unit tests will fail. [Ignoring tests in specific platforms is not yet possible](https://youtrack.jetbrains.com/issue/KT-23584), but it should be possible to use the `expect` and `actual` trick again to work around this.
